packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs

```diff
@@ -11,38 +13,76 @@ namespace Microsoft.Generator.CSharp.Providers
         public MethodSignatureBase Signature { get; }
         public MethodBodyStatement? BodyStatements { get; }
         public ValueExpression? BodyExpression { get; }
+        public XmlDocProvider XmlDocs { get; }
 
         /// <summary>
         /// Initializes a new instance of the <see cref="MethodProvider"/> class with a body statement and method signature.
         /// </summary>
         /// <param name="signature">The method signature.</param>
         /// <param name="bodyStatements">The method body.</param>
-        /// <param name="kind">The method kind <see cref="CSharpMethodKinds"/>.</param>
-        public MethodProvider(MethodSignatureBase signature, MethodBodyStatement bodyStatements, CSharpMethodKinds? kind = null)
+        /// <param name="skipParamValidation">Indicates if validation statements for parameters should be added.</param>
+        public MethodProvider(MethodSignatureBase signature, MethodBodyStatement bodyStatements, bool skipParamValidation = false)
         {
             Signature = signature;
-            BodyStatements = bodyStatements;
-            Kind = kind ?? new CSharpMethodKinds("Undefined");
+            skipParamValidation = skipParamValidation || !signature.Modifiers.HasFlag(MethodSignatureModifiers.Public);
+            List<MethodBodyStatement> statements = skipParamValidation ? new List<MethodBodyStatement>() : [.. GetValidationStatements()];
+            statements.Add(bodyStatements);
+            BodyStatements = statements;
+            XmlDocs = BuildXmlDocs(skipParamValidation);
         }
 
         /// <summary>
         /// Initializes a new instance of the <see cref="MethodProvider"/> class with a body expression and method signature.
         /// </summary>
         /// <param name="signature">The method signature.</param>
         /// <param name="bodyExpression">The method body expression.</param>
-        /// <param name="kind">The method kind <see cref="CSharpMethodKinds"/>.</param>
-        public MethodProvider(MethodSignatureBase signature, ValueExpression bodyExpression, CSharpMethodKinds? kind = null)
+        public MethodProvider(MethodSignatureBase signature, ValueExpression bodyExpression)
         {
             Signature = signature;
             BodyExpression = bodyExpression;
-            Kind = kind ?? new CSharpMethodKinds("Undefined");
+            XmlDocs = BuildXmlDocs(true);
+        }
+
+
+        private IEnumerable<MethodBodyStatement> GetValidationStatements()
+        {
+            bool wroteValidation = false;
+            foreach (var parameter in Signature.Parameters)
+            {
+                if (parameter.Validation != ParameterValidationType.None)
+                {
+                    yield return Argument.ValidateParameter(parameter);
+                    wroteValidation = true;
+                }
+            }
+            if (wroteValidation)
+                yield return EmptyLineStatement;
+        }
```

Comment:
I think maybe we could change this to return a `IReadOnlyList<>` and construct it using a `List` to improve the performance because we are trying to avoid using yield return because of performance issues in other places.

Reply:
Good catch, when I looked at this again I am already creating a hash of parameters that need validation in the xml doc builder so I moved that out so I can use the hash in both places and then I can construct an array of the exact size needed and that seems to be the best perf.

```
|        Method |      Mean |    Error |   StdDev |   Gen0 | Allocated |
|-------------- |----------:|---------:|---------:|-------:|----------:|
|   Yieldreturn | 124.90 ns | 2.125 ns | 1.988 ns | 0.0251 |     424 B |
|       UseList | 137.69 ns | 1.360 ns | 1.205 ns | 0.0325 |     544 B |
| UseSingleList |  93.68 ns | 0.748 ns | 0.625 ns | 0.0236 |     400 B |
|      UseArray |  74.51 ns | 1.076 ns | 0.898 ns | 0.0214 |     360 B |
```